### PR TITLE
[CPP Graph] add s8 perchannel quant and kernel.

### PR DESCRIPTION
**Pull request by luoyu-intel**
## Type of Change

1. add s8 per channel quantitation and kernel
2. re-org jblas template file
3. add back comp_type=fp32 kernel
4. add comp_type=bf16 
5. add workspace pre-allocation

## How has this PR been tested?

GPTJ    comp_type=[int8,fp32,bf16], bits=[4,8], group_size=[-1,128], 
group_size=-1 only for int8 per channel.

### CODE REVIEW

**Comment by luoyu-intel** [comment]
perchannel-q8:
./bin/quant_gptj --model_file ~/luoyu/gpt-f16.bin --out_file ./gptj_q8_j_int8_fp32_pern.bin --bits 8 --alg sym --block_size -1 --scale_dtype fp32 --compute_type int8 --model_name gptj

bf16-q4:
./bin/quant_gptj --model_file ~/luoyu/gpt-f16.bin --out_file ./gptj_q4_j_bf16_fp32_b128.bin --bits 4 --alg sym --block_size 128 --scale_dtype fp32 --compute_type bf16 --model_name gptj

fp32-q4:
./bin/quant_gptj --model_file ~/luoyu/gpt-f16.bin --out_file ./gptj_q4_j_fp32_fp32_b128.bin --bits 4 --alg sym --block_size 128 --scale_dtype fp32 --compute_type fp32 --model_name gptj

**Comment by hshen14** [comment]
> perchannel-q8: ./bin/quant_gptj --model_file ~/luoyu/gpt-f16.bin --out_file ./gptj_q8_j_int8_fp32_pern.bin --bits 8 --alg sym --block_size -1 --scale_dtype fp32 --compute_type int8 --model_name gptj
> 
> bf16-q4: ./bin/quant_gptj --model_file ~/luoyu/gpt-f16.bin --out_file ./gptj_q4_j_bf16_fp32_b128.bin --bits 4 --alg sym --block_size 128 --scale_dtype fp32 --compute_type bf16 --model_name gptj
> 
> fp32-q4: ./bin/quant_gptj --model_file ~/luoyu/gpt-f16.bin --out_file ./gptj_q4_j_fp32_fp32_b128.bin --bits 4 --alg sym --block_size 128 --scale_dtype fp32 --compute_type fp32 --model_name gptj

as we discussed, shall we rename block to group in the context of INT4 quantization?

**Comment by luoyu-intel** [comment]
> as we discussed, shall we rename block to group in the context of INT4 quantization?
it's okay to rename the parameter name. block is the continuation of llama.cpp's definition.

File: intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_weight_compression.h [hunk 1]
```diff
@@ -1292,13 +1378,13 @@ using GemmSKernelDynamicS4FullRangeFp32KBlock = jblas::wrapper::gemm_kblock::Gem
         jblas::epilogue::gemm::AccumulatorWriteBackFp32>,
     jblas::utils::parallel::Parallel2DGemmKBlockFixed>;
 
-using GemmDynamicS4Fp32PerChannelN = jblas::wrapper::gemm_pack_weight::GemmInterfaceAB<
+using GemmDynamicS8Fp32PerChannelN = jblas::wrapper::gemm_pack_weight::GemmInterfaceParallelAB<
     jblas::wrapper::gemm_pack_weight::GemmLauncherPackWeight<
-        DefaultISA, jblas::gemm::GemmCore_Row_NN_16x48_AMX_INT8_ss,
+        DefaultISA, jblas::gemm::GemmCore_Row_NN_16x48_AMX_S8S8,  //
```
Comment (zhewang1-intc) [inline intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_weight_compression.h:1383]: Is this gemmcore's coretype==AMX_INT8_16x48_SS?

File: intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_epilogue.h [hunk 2]
```diff
@@ -35,20 +39,24 @@ class AccumulatorWriteBack {
                      const int N, const Param& _param, Eltops... ops) {
     auto COffset = M_offset * _param.ldc + N_offset;
     auto cptr = _param.C + COffset;
-    bool constexpr Valid = !std::is_same<_DST_T, jblas::utils::bf16>::value ? true : std::is_same<_SRC_T, float>::value;
+    bool constexpr Valid = !std::is_same<DType, utils::bf16>::value || std::is_same<SType, float>::value;
     static_assert(Valid, "fp32 to bf16 conversion only.");
-    if (std::is_same<_DST_T, jblas::utils::bf16>::value) {
+    if (std::is_same<DType, utils::bf16>::value) {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_epilogue.h:44]: should we add constexpr before any condition judge in templates

File: intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_weight_compression.h (hunk 1 above)
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_weight_compression.h:1383]: AMX_INT8_16x48 is okay. For now, only NTILE matters. After the support of asymmetric weight, we need to split it into two types.

File: intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_weight_compression.h [hunk 3]
```diff
@@ -1292,13 +1378,13 @@ using GemmSKernelDynamicS4FullRangeFp32KBlock = jblas::wrapper::gemm_kblock::Gem
         jblas::epilogue::gemm::AccumulatorWriteBackFp32>,
     jblas::utils::parallel::Parallel2DGemmKBlockFixed>;
 
-using GemmDynamicS4Fp32PerChannelN = jblas::wrapper::gemm_pack_weight::GemmInterfaceAB<
+using GemmDynamicS8Fp32PerChannelN = jblas::wrapper::gemm_pack_weight::GemmInterfaceParallelAB<
```
Comment (zhewang1-intc) [inline intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_weight_compression.h:1381]: btw, i replaced all `GemmInterfacePackWeight` with `GemmInterfaceParallelAB` in weight_compression kernel yesterday, but failed to compile. Just a reminder.

Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_weight_compression.h:1381]: ok

File: intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_epilogue.h (hunk 2 above)
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_epilogue.h:44]: yes, we should add constexpr as a standard behavior

File: intel_extension_for_transformers/llm/runtime/graph/core/jblas_common.hpp [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+constexpr jblas::gemm::GemmCoreType GcCompInt8KBlockSet[] = {jblas::gemm::GemmCoreType::AMX_INT8_16X48_KBLOCK,
+                                                             jblas::gemm::GemmCoreType::AVX512_VNNI_3X48_KBLOCK};
+
+constexpr jblas::gemm::GemmCoreType GcCompInt8Set[] = {jblas::gemm::GemmCoreType::AMX_INT8_16x48_SS,
+                                                       jblas::gemm::GemmCoreType::AVX512_VNNI_8X48};
+
+namespace custom {
+namespace epilogue {
+template <JBLAS_ISA ISA_T, typename _T>
+class Silu {
```
[the remark this answers is not in the document]
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/runtime/graph/core/jblas_common.hpp:122]: A good point.

[the remark this answers is not in the document]
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/runtime/graph/core/jblas_common.hpp:122]: move jblas_common to layers folder.

**Review by airMeng — APPROVED** [review]

**Review by zhewang1-intc — APPROVED** [review]
LGTM

**Pull request merged**